sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,206 @@
+
+                return refreshMono.onErrorResume(
+                    (exception) -> {
+                        if (logger.isDebugEnabled()) {
+                            logger.debug("refresh cache [{}] resulted in error", key, exception);
+                        }
+
+                        if (initialLazyValue.shouldRemoveFromCache()) {
+                            // In some scenarios when a background failure occurs like a 404 the initial cache value should be removed.
+                            if (removeNotFoundFromCacheException((CosmosException)exception)) {
+                                this.remove(key);
+                            }
+                        }
+                        return Mono.error(exception);
+                    }
+                );
+            }).onErrorResume((exception) -> {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("cache[{}] resulted in error", key, exception);
+                }
+                if (initialLazyValue.shouldRemoveFromCache()) {
+                    this.remove(key);
+                }
+                return Mono.error(exception);
+            });
+        }
+
+        if (logger.isDebugEnabled()) {
+            logger.debug("cache[{}] doesn't exist, computing new value", key);
+        }
+        AsyncLazyWithRefresh<TValue> asyncLazyWithRefresh = new AsyncLazyWithRefresh<TValue>(singleValueInitFunc);
+        this.values.putIfAbsent(key, asyncLazyWithRefresh);
+        AsyncLazyWithRefresh<TValue> result = this.values.get(key);
+
+        return result.getValueAsync().onErrorResume(
+            (exception) -> {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("cache[{}] resulted in error", key, exception);
+                }
+                // Remove the failed task from the dictionary so future requests can send other calls..
+                if (result.shouldRemoveFromCache()) {
+                    this.remove(key);
+                }
+                return Mono.error(exception);
+            }
+        );
+    }
+
+    public void set(TKey key, TValue value) {
+        if (logger.isDebugEnabled()) {
+            logger.debug("set cache[{}]={}", key, value);
+        }
+        AsyncLazyWithRefresh<TValue> updatedValue = new AsyncLazyWithRefresh<TValue>(value);
+        this.values.put(key, updatedValue);
+    }
+
+    public void remove(TKey key) {
+        values.remove(key);
+    }
+
+    /**
+     * This is AsyncLazy that has an additional Task that can
+     * be used to update the value. This allows concurrent requests
+     * to use the stale value while the refresh is occurring.
+     */
+    private class AsyncLazyWithRefresh<TValue> {
+//        private final Function<TValue, Mono<TValue>> createValueFunc;
+        private final AtomicBoolean removeFromCache = new AtomicBoolean(false);
+        private AtomicReference<Mono<TValue>> value;
+        private Mono<TValue> refreshInProgress;
```

Comment:
I didn't understand why we need `value` and `refreshInProgress`. I think only `value` should be kept and `refreshInProgress` should be removed. Given `refreshInProgress` is null by default and it will only be non-null when we apply the refresh function, which is how at first place `value` is updated in the constructor.

Reply:
As discussed offline, updated the code for the corner cases.